src/common/comp/brute_force.cc

```diff
@@ -786,84 +786,85 @@ BruteForce::AnnIterator(const DataSetPtr base_dataset, const DataSetPtr query_da
+                                                  distances_ids, id_selector);
+                            } else if constexpr (KnowhereHalfPrecisionFloatPointTypeCheck<DataType>::value) {
+                                // normalize query vector may cause presision loss, so div query norms in apply function
+
+                                faiss::half_precision_floating_point_all_cosine(cur_query, (const DataType*)xb,
+                                                                                norms.get(), dim, 1, nb, distances_ids,
+                                                                                id_selector);
+                            } else {
+                                LOG_KNOWHERE_ERROR_ << "Metric Cosine only used in floating point data";
+                                throw std::runtime_error("Metric Cosine only used in floating point data");
+                            }
                         } else {
-                            LOG_KNOWHERE_ERROR_ << "Metric Cosine only used in floating point data";
-                            return Status::faiss_inner_error;
+                            if constexpr (std::is_same_v<DataType, knowhere::fp32>) {
+                                faiss::all_inner_product(cur_query, (const float*)xb, dim, 1, nb, distances_ids,
+                                                         id_selector);
+                            } else if constexpr (KnowhereHalfPrecisionFloatPointTypeCheck<DataType>::value) {
+                                faiss::half_precision_floating_point_all_inner_product(
+                                    cur_query, (const DataType*)xb, dim, 1, nb, distances_ids, id_selector);
+                            } else {
+                                LOG_KNOWHERE_ERROR_ << "Metric Cosine only used in floating point data";
+                                throw std::runtime_error("Metric Cosine only used in floating point data");
+                            }
                         }
+                        break;
+                    }
+                    default: {
+                        LOG_KNOWHERE_ERROR_ << "Invalid metric type: " << cfg.metric_type.value();
+                        throw std::runtime_error("Invalid metric type");
                     }
-                    break;
-                }
-                default: {
-                    LOG_KNOWHERE_ERROR_ << "Invalid metric type: " << cfg.metric_type.value();
-                    return Status::invalid_metric_type;
                 }
-            }
-            if (xb_id_offset != 0) {
-                for (auto& distances_id : distances_ids) {
-                    distances_id.id = distances_id.id == -1 ? -1 : distances_id.id + xb_id_offset;
+                if (xb_id_offset != 0) {
+                    for (auto& distances_id : distances_ids) {
+                        distances_id.id = distances_id.id == -1 ? -1 : distances_id.id + xb_id_offset;
```

Comment:
just to confirm: this is evaluated as `distances_id.id = a ? b : c`, not `distances_id.id = (a ? b : c) + xb_id_offset'?

Reply:
@cqy123456 help check~

Reply:
yes, `-1`  or `distances_id.id + xb_id_offset`